### DISCUSSION TITLE
Integration with time tracker website/addon

### DISCUSSION THREAD

**Discussion by glaebhoerl** [proginosko/LeechBlockNG#474] [Ideas]
I use [Clockify](clockify.me) (together with its [Firefox addon](https://clockify.me/firefox-time-tracking)) to track the time spent e.g. working on a consulting project. It'd be amazing if I could somehow set things up so that LeechBlock only lets me access *certain websites* once I've logged a given number of hours that day in Clockify.

I assume there are many other similar services, so presumably, if this were added as an upstream feature, it would have to be in a more general form to potentially support multiple providers? (Which, alas, sounds like a large project.)

In any case, this would be valuable enough to me that I'm definitely motivated to get something working, even if it's just a pile of hacks in a perpetually local fork. If you have any advice on what the least painful implementation strategy might be, I would very much appreciate it!

**Comment by proginosko** [comment]
Does Clockify have an API or some other way for third-party apps to access time logged?

**Reply by glaebhoerl** [reply]
It does: https://docs.clockify.me/

Here's part of a shell script I threw together a while ago to print summaries of the time logged today resp. this month, for reference:
<details>

```bash
function get() {
    echo "$(curl --silent --header "X-Api-Key: $API_KEY" "$@")"
}

function get_user_id() {
    get "https://api.clockify.me/api/v1/user" | jq -cr '.id'
}

function get_workspaces() {
    get "https://api.clockify.me/api/v1/workspaces" | jq -cr '.[] | "\(.name):\(.id)"'
}

function get_time_from_to() {
    local _USER="$1"; shift
    local _WORKSPACE="$1"; shift
    local _BEGIN="$1"; shift
    local _END="$1"; shift
    get --request POST --header 'content-type: application/json' "https://reports.api.clockify.me/v1/workspaces/$_WORKSPACE/reports/summary" \
        --data '{"summaryFilter": {"groups": ["USER"]}, "dateRangeStart": "'$_BEGIN'", "dateRangeEnd": "'$_END'", "users":{"ids":["'$_USER'"],"contains":"CONTAINS","status":"ALL"}, "amountShown": "HIDE_AMOUNT"}' | jq ".totals[0].totalTime"
}

function get_time_this_month() {
    local _MONTH=$(date "+%Y-%m")
    local _NEXTMONTH=$(date "+%Y-%m" --date="next month")
    local _BEGIN="$_MONTH-01T00:00:00.000Z"
    local _END="$_NEXTMONTH-01T00:00:00.000Z"
    get_time_from_to $1 $2 $_BEGIN $_END
}

function get_time_today() {
    local _HOUR=$(date +"%-H")

    local _TODAY=$(date "+%Y-%m-%d")
    if ((_HOUR < 6)); then
        _TODAY=$(date "+%Y-%m-%d" --date="yesterday")
    fi

    local _TOMORROW=$(date "+%Y-%m-%d" --date="tomorrow")
    local _BEGIN="${_TODAY}T06:00:00.000Z"
    local _END="${_TOMORROW}T06:00:00.000Z"
    get_time_from_to $1 $2 $_BEGIN $_END
}

function get_time_in_progress() {
    local _USER=$1; shift
    local _WORKSPACE=$1; shift
    local _DATE=$(get "https://api.clockify.me/api/v1/workspaces/$_WORKSPACE/user/$_USER/time-entries?in-progress=true" | jq -cr '.[0].timeInterval.start')
    if [[ "$_DATE" == "null" ]]; then
        echo 0
        return
    fi
    echo $(( $(date +"%s") - $(date +"%s" --date="$_DATE") ))
}
```
</details>
